project/src/hooks/useMap/useMap.tsx

```diff
@@ -0,0 +1,36 @@
+import { useEffect, useState, MutableRefObject } from 'react';
+import { instanceLayer, layerAttribution } from '../../const/map';
+import { Map, TileLayer } from 'leaflet';
+import { City } from '../../types/offer';
+
+export default function useMap(mapRef: MutableRefObject<HTMLElement | null>, city: City): Map | null {
+  const [map, setMap] = useState<Map | null>(null);
+  const { latitude, longitude, zoom } = city.location;
+
+  useEffect(() => {
+    if (mapRef.current !== null && map === null) {
+      const instance = new Map(mapRef.current, {
```

Comment:
а что касается сохранения инстанса, ты сейчас сохраняешь в map, то есть используешь useState, можно использовать для этого реф, то есть что то типа такого
```suggestion
  const map = useRef<Map | null>(null);
  const { latitude, longitude, zoom } = city.location;

  useEffect(() => {
    if (mapRef.current !== null && map === null) {
      const instance = new Map(mapRef.current, {
      .....
      map.current = instance;
      
```
это синхронная операция, и даже если компонент перерисуется сразу после этого всё должно быть хорошо

Reply:
ну либо есть ещё вариант с флагом, когда в реф записывают не сам инстанс, а логическое значение, и после первого рендера меняют его на противоположное значение, и потом смотрят перед созданием инстанса чему оно равно